src/lightning/fabric/utilities/throughput_monitor.py

```diff
@@ -0,0 +1,354 @@
+    | `items_per_sec`          | Rolling average (over `window_size` most recent batches) of the number of items       |
+    |                          | processed per second                                                                  |
+    +--------------------------+---------------------------------------------------------------------------------------+
+    | `flops_per_sec`          | Estimates flops by `flops_per_batch * batches_per_sec`                                |
+    +--------------------------+---------------------------------------------------------------------------------------+
+    | `device/batches_per_sec` | `batches_per_sec` divided by world size                                               |
+    +--------------------------+---------------------------------------------------------------------------------------+
+    | `device/samples_per_sec` | `samples_per_sec` divided by world size                                               |
+    +--------------------------+---------------------------------------------------------------------------------------+
+    | `device/items_per_sec`   | `items_per_sec` divided by world size. This may include padding depending on the data |
+    +--------------------------+---------------------------------------------------------------------------------------+
+    | `device/flops_per_sec`   | `flops_per_sec` divided by world size.                                                |
+    +--------------------------+---------------------------------------------------------------------------------------+
+    | `device/mfu`             | `device/flops_per_sec` divided by world size.                                         |
+    +--------------------------+---------------------------------------------------------------------------------------+
+    | `time/train`             | Total elapsed training time                                                           |
+    +--------------------------+---------------------------------------------------------------------------------------+
+    | `time/val`               | Total elapsed validation time                                                         |
+    +--------------------------+---------------------------------------------------------------------------------------+
+    | `time/total`             | Total elapsed time (time/train + time/val)                                            |
+    +--------------------------+---------------------------------------------------------------------------------------+
+
+    Notes:
+        - The implementation assumes that devices are homogeneous as it normalizes by the world size.
+        - items_per_sec, flops_per_sec and MFU do not account for padding if present. We suggest using
+            samples_per_sec or batches_per_sec to measure throughput under this circumstance.
+
+    Args:
+        fabric: The fabric object.
```

Comment:
```suggestion
        fabric: The Fabric object.
```

Reply:
If this is treated like a "callback", wouldn't it be much better if we could pass the fabric and model objects to the hook instead of the init? Because otherwise there is a circular dependency: You can't create the callback unless you have the fabric object, and you can create the fabric object to pass in the callback because you can't create it yet.

My suggestion here is either we go all in on the callback thing, or we make it clear it is NOT a callback but a utility instead.

Reply:
I'm confused by your second comment. This is not a callback (it's only in the Trainer) but and utility class that you interact with. See the added test https://github.com/Lightning-AI/lightning/pull/18848/files#diff-3e8b95020b6ac7ba6c95cc2be30730b9dec24d9de89139b39ee28b453d9bb731R29-R52 as an example

Reply:
Yeah, in Fabric anything can be a callback. That system is anyway the weirdest concept in there, but my point was that the the thing looks in every way like a callback the way it is designed now, so you could (not saying you should) use it with `Fabric(callbacks=ThroughputMonitor())` and it would work. As long as we don't mention anywhere that it is a callback, I'm ok with it.

Reply:
I did some changes in this direction, see the comment after https://github.com/Lightning-AI/lightning/pull/18848/files/202d764502f4f1c41ac9af190fd64236f6068076#r1372135201